src/sql/src/plan/statement/ddl.rs

```diff
@@ -2995,97 +2995,205 @@ fn kafka_sink_builder(
+            avro_value_fullname,
+            set_null_defaults: null_defaults,
+            is_debezium: matches!(envelope, SinkEnvelope::Debezium),
+            sink_from: Some(sink_from),
+            value_doc_options,
+            key_doc_options,
+        };
+        Ok((
+            csr_connection,
+            options,
+            key_compatibility_level,
+            value_compatibility_level,
+        ))
+    };
 
-            let options = AvroSchemaOptions {
-                avro_key_fullname,
-                avro_value_fullname,
-                set_null_defaults: null_defaults,
-                is_debezium: matches!(envelope, SinkEnvelope::Debezium),
-                sink_from: Some(sink_from),
-                value_doc_options,
-                key_doc_options,
-            };
+    // Helper function to determine if a relation desc is valid for text or binary format.
+    // We only allow a single column of non-list/map type.
+    let valid_text_binary_desc = |desc: &RelationDesc| {
+        if desc.arity() != 1 {
+            false
+        } else if desc.iter().any(|(_, ty)| {
+            matches!(
+                ty.scalar_type,
+                ScalarType::List { .. } | ScalarType::Map { .. }
+            )
+        }) {
+            false
+        } else {
+            true
+        }
+    };
 
+    let map_format_non_avro = |format: &Format<Aug>, desc: &RelationDesc| match format {
+        Format::Json { array: false } => Ok::<_, PlanError>(KafkaSinkFormatType::Json),
+        Format::Bytes if valid_text_binary_desc(desc) => Ok(KafkaSinkFormatType::Bytes),
+        Format::Text if valid_text_binary_desc(desc) => Ok(KafkaSinkFormatType::Text),
+        Format::Bytes | Format::Text => {
+            bail_unsupported!("BYTE/TEXT format with multiple columns or non-scalar types")
+        }
+        Format::Json { array: true } => bail_unsupported!("JSON ARRAY format in sinks"),
+        format => bail_unsupported!(format!("sink format {:?}", format)),
+    };
+
+    // Map from the format specifier of the statement to the individual key/value formats for the sink.
+    let format = match format {
+        Some(FormatSpecifier::KeyValue {
+            key: Format::Avro(..),
+            value: Format::Avro(..),
+        }) => {
+            // Since Avro format requires specifying CSR connection options, if we allow both key and value
+            // to be separately specified as Avro we would need to deconflict the CSR connection options
+            // between them, so we disallow this.
+            sql_bail!("Use FORMAT AVRO instead of KEY FORMAT AVRO..VALUE FORMAT AVRO..");
+        }
+        Some(FormatSpecifier::KeyValue {
+            key: Format::Avro(..),
+            value,
+        }) => {
+            // Supporting key as avro and value as non-avro is tricky without completely refactoring
+            // the AvroSchemaGenerator, so disable it since it is unlikely to be useful anways.
```

Comment:
Which part of `AvroSchemaGenerator` needs the key and value shcemas to exist at the same time? Its new method seems to be easily splittable into two. I see the key being used at the tail end and doesn't use any values derived by the value schema.

Reply:
you're right - I'll refactor `AvroSchemaGenerator` to split it into two and then we can support this case